codyze-core/src/main/kotlin/de/fraunhofer/aisec/codyze/Sarif.kt

```diff
@@ -0,0 +1,216 @@
+    return Message(text = "${this?.javaClass?.simpleName}[name=${this?.name}]")
+}
+
+/**
+ * Converts a [Node] into a [Stack] for SARIF output. Currently, this is a single stack frame with
+ * the location of the node. In the future, we want to include a call stack that leads to the node's
+ * current function.
+ */
+private fun Node.toSarifCallStack(): List<Stack> {
+    val currentFunc = this.firstParentOrNull { it is FunctionDeclaration }
+    return listOf(
+        Stack(
+            message = Message(text = "Stack"),
+            frames =
+                listOf(
+                    StackFrame(
+                        location = this.toSarifLocation(message = currentFunc.toSarifMessage())
+                    )
+                ),
+        )
+    )
+}
+
+/** Converts a list of [Node]s into a list of [ThreadFlowLocation]s for SARIF output. */
+private fun List<Node>?.toSarifThreadFlowLocation(): List<ThreadFlowLocation> {
+    return this?.mapIndexed { idx, node ->
+        ThreadFlowLocation(
+            location =
+                node.toSarifLocation(message = node.toSarifMessage(), onlyFunctionHeader = true),
+            executionOrder = idx.toLong(),
+        )
+    } ?: listOf()
+}
+
+/** Converts a [Node.location] to a [Location]. */
+fun Node?.toSarifLocation(
+    message: Message? = this.toSarifMessage(),
+    /**
+     * If this option is set to true, we only emit the location of the function header, not the
+     * entire function body.
+     *
+     * This is helpful for cases where we want to highlight the function declaration in the code
+     * editor, but not the entire function body.
+     */
+    onlyFunctionHeader: Boolean = false,
+): Location? {
+    val location = this?.location ?: return null
+
+    return if (this is FunctionDeclaration && this.body != null && onlyFunctionHeader) {
+            // Try to calculate the end of the header by using the beginning of the body. This is
+            // not entirely correct since in some programming languages we need to start the body
+            // location at the first statement, since we are missing location information for the
+            // body "block", but it's the best we can do
+            val location =
+                de.fraunhofer.aisec.cpg.sarif.PhysicalLocation(
+                    uri = location.artifactLocation.uri,
+                    region =
+                        de.fraunhofer.aisec.cpg.sarif.Region(
+                            startLine = location.region.startLine,
+                            startColumn = location.region.startColumn,
+                            endLine =
+                                this.body?.location?.region?.startLine ?: location.region.endLine,
+                            endColumn =
+                                this.body?.location?.region?.startColumn
+                                    ?: location.region.endColumn,
+                        ),
+                )
+            location
```

Comment:
We're returning the value of the last expression. We can simply call the constructor of `PhysicalLocation`.